File: tmdt-laravel/app/Http/Controllers/AdminController.php
<?php

namespace App\Http\Controllers;

use Illuminate\Http\Request;
use App\Models\NguoiDung;
use App\Models\SanPham;
use App\Models\LoaiSanPham;
use App\Models\HoaDon;
use App\Models\KhieuNai;
use App\Models\HinhAnhSP;
use Illuminate\Support\Facades\Session;

class AdminController extends Controller
{
    private function checkAdmin()
    {
        $user = Session::get('user');
        if (!$user || $user->VaiTro != 'Admin') {
            abort(403, 'Unauthorized action.');
        }
    }

    public function index()
    {
        $this->checkAdmin();
        $tongNguoiDung = NguoiDung::count();
        $tongSanPham = SanPham::count();
        $daban = SanPham::where('TrangThai', 'Đã bán')->count();
        $an = SanPham::where('TrangThai', 'Ẩn')->count();
        $tinDaDuyet = SanPham::where('TrangThai', 'Đã duyệt')->count();

        return view('admin.index', compact('tongNguoiDung', 'tongSanPham', 'daban', 'an', 'tinDaDuyet'));
    }

    public function doiTrangThai(Request $request)
    {
        $this->checkAdmin();
        $id = $request->input('id');
        $sp = SanPham::find($id);
        if (!$sp) abort(404);

        $sp->TrangThai = $request->input('tt');
        $sp->save();

        return redirect()->route('admin.quanlysanpham')->with('success', "✅ Đã cập nhật trạng thái sản phẩm **{$sp->TenSP}** thành '{$request->input('tt')}'");
    }

    public function quanLyNguoiDung()
    {
        $this->checkAdmin();
        $dsNguoiDung = NguoiDung::orderBy('MaKH', 'desc')->get();
        return view('admin.quanlynguoidung', compact('dsNguoiDung'));
    }

    public function quanLySanPham()
    {
        $this->checkAdmin();
        $SanPhams = SanPham::with('nguoiDung')->orderBy('MaSP', 'desc')->get();
        $LoaiSanPhams = LoaiSanPham::with('sanPhams')->get();

        return view('admin.quanlysanpham', compact('SanPhams', 'LoaiSanPhams'));
    }

    public function quanLyDonHang()
    {
        $this->checkAdmin();
        $donhangs = HoaDon::with(['nguoiDung', 'ctHoaDons.sanPham.nguoiDung'])
            ->get()
            ->map(function ($hd) {
                $nguoiBan = "(Chưa có sản phẩm)";
                if ($hd->ctHoaDons->isNotEmpty() && $hd->ctHoaDons->first()->sanPham && $hd->ctHoaDons->first()->sanPham->nguoiDung) {
                    $nguoiBan = $hd->ctHoaDons->first()->sanPham->nguoiDung->HoTen;
                }
                return [
                    'MaHD' => $hd->MaHD,
                    'NguoiMua' => $hd->nguoiDung->HoTen ?? '',
                    'NguoiBan' => $nguoiBan,
                    'NgayDat' => $hd->NgayDat,
                    'TongTien' => $hd->TongTien,
                    'TrangThai' => $hd->TrangThai
                ];
            });

        return view('admin.quanlydonhang', compact('donhangs'));
    }

    public function quanLyKhieuNai()
    {
        $this->checkAdmin();
        $dsKhieuNai = KhieuNai::with(['nguoiDung', 'sanPham'])->orderBy('NgayGui', 'desc')->get();
        return view('admin.quanlykhieunai', compact('dsKhieuNai'));
    }

    public function capNhatTrangThaiKN(Request $request)
    {
        $user = Session::get('user');
        if (!$user || $user->VaiTro != 'Admin') abort(403);

        $id = $request->input('id');
        $kn = KhieuNai::find($id);
        if (!$kn) abort(404);

        $kn->TrangThai = "Đã giải quyết";
        // There is no PhanHoi column in KhieuNai migration we made, so skip or update if added later.
        $kn->save();

        return redirect()->route('admin.quanlykhieunai');
    }

    public function xoaKhieuNai(Request $request)
    {
        $user = Session::get('user');
        if (!$user || $user->VaiTro != 'Admin') abort(403);

        $id = $request->input('id');
        $kn = KhieuNai::find($id);
        if (!$kn) abort(404);

        $kn->delete();
        return redirect()->route('admin.quanlykhieunai')->with('success', 'Đã xoá khiếu nại thành công!');
    }

    public function xoa(Request $request)
    {
        $this->checkAdmin();
        
        $id = $request->input('id');
        $sp = SanPham::find($id);
        if (!$sp) return redirect()->route('admin.quanlysanpham')->with('error', 'Sản phẩm không tồn tại!');

        try {
            $hinhAnhs = HinhAnhSP::where('MaSP', $id)->get();
            $path = public_path('Content/Images/');
            foreach ($hinhAnhs as $item) {
                if ($item->URLAnh != 'noimage.jpg' && !str_starts_with($item->URLAnh, 'http')) {
                    @unlink($path . $item->URLAnh);
                }
            }
            HinhAnhSP::where('MaSP', $id)->delete();
            $sp->delete();
            return redirect()->route('admin.quanlysanpham')->with('success', '🗑️ Đã xóa sản phẩm vĩnh viễn!');
        } catch (\Exception $e) {
            $sp->TrangThai = 'Ẩn';
            $sp->save();
            return redirect()->route('admin.quanlysanpham')->with('success', '⚠️ Sản phẩm đã có đơn hàng, không thể xóa hẳn. Hệ thống đã chuyển sang trạng thái "Ẩn".');
        }
    }

    public function doiTrangThaiNguoiDung(Request $request)
    {
        $this->checkAdmin();
        $id = $request->input('id');
        $user = NguoiDung::find($id);
        if (!$user) return redirect()->route('admin.quanlynguoidung')->with('error', 'Không tìm thấy tài khoản!');

        if ($user->VaiTro == 'Admin') {
            return redirect()->route('admin.quanlynguoidung')->with('error', 'Không thể khóa tài khoản Admin!');
        }

        $user->Khoa = !$user->Khoa;
        $user->save();

        $msg = $user->Khoa ? 'Tài khoản đã bị khóa!' : 'Tài khoản đã được mở khóa!';
        return redirect()->route('admin.quanlynguoidung')->with('success', $msg);
    }
}

File: legacy/aspnet-mvc/ThuongMaiDienTu-DoAn/Controllers/AdminController.cs
//using System;
//using System.Data.Entity;
//using System.IO;
//using System.Linq;
//using System.Web.Mvc;
//using ThuongMaiDienTu_DoAn.Filters;
//using ThuongMaiDienTu_DoAn.Models;

//namespace ThuongMaiDienTu_DoAn.Controllers
//{
//    [AuthorizeAdmin]
//    public class AdminController : Controller
//    {
//        private readonly TMDTEntities db = new TMDTEntities();

//        // === DASHBOARD ===
//        public ActionResult Index()
//        {
//            ViewBag.TongNguoiDung = db.NGUOIDUNGs.Count();
//            ViewBag.TongSanPham = db.SANPHAMs.Count();
//            ViewBag.TinChoDuyet = db.SANPHAMs.Count(sp => sp.TrangThai == "Chờ duyệt");
//            ViewBag.TinDaDuyet = db.SANPHAMs.Count(sp => sp.TrangThai == "Đã duyệt");
//            return View();
//        }

//        // === DUYỆT TIN / SẢN PHẨM ===


//        [HttpPost]
//        public ActionResult DoiTrangThai(int id, string tt)
//        {
//            var sp = db.SANPHAMs.Find(id);
//            if (sp == null)
//                return HttpNotFound();

//            sp.TrangThai = tt;
//            db.SaveChanges();

//            TempData["Success"] = $"✅ Đã cập nhật trạng thái sản phẩm **{sp.TenSP}** thành '{tt}'";
//            return RedirectToAction("QuanLySanPham");
//        }

//        // === QUẢN LÝ NGƯỜI DÙNG ===
//        public ActionResult QuanLyNguoiDung()
//        {
//            var ds = db.NGUOIDUNGs.OrderByDescending(x => x.NgayTao).ToList();
//            return View(ds);
//        }

//        // === QUẢN LÝ SẢN PHẨM ===
//        public ActionResult QuanLySanPham()
//        {
//            // 1. Lấy dữ liệu Sản phẩm
//            var listSP = db.SANPHAMs
//                .Include(s => s.NGUOIDUNG)
//                .OrderByDescending(x => x.NgayTao)
//                .ToList();

//            // 2. Lấy dữ liệu Loại (kèm sản phẩm con)
//            var listLoai = db.LOAISANPHAMs.Include("SANPHAMs").ToList();

//            // 3. Đổ vào ViewModel (lúc này nó tự hiểu class nằm trong thư mục Models)
//            var model = new AdminProductViewModel
//            {
//                SanPhams = listSP,
//                LoaiSanPhams = listLoai
//            };

//            return View(model);
//        }

//        // === QUẢN LÝ ĐƠN HÀNG ===
//        public ActionResult QuanLyDonHang()
//        {
//            var donhangs = db.HOADONs
//                .Select(hd => new DonHangC2CViewModel
//                {
//                    MaHD = hd.MaHD,
//                    NguoiMua = hd.NGUOIDUNG.HoTen,
//                    NguoiBan = hd.CT_HOADON
//                                 .Select(ct => ct.SANPHAM.NGUOIDUNG.HoTen)
//                                 .FirstOrDefault() ?? "(Chưa có sản phẩm)",
//                    NgayDat = hd.NgayDat,
//                    TongTien = (decimal)hd.TongTien,
//                    TrangThai = hd.TrangThai
//                })
//                .ToList();

//            return View(donhangs);
//        }

//        // === QUẢN LÝ KHIẾU NẠI ===
//        public ActionResult QuanLyKhieuNai()
//        {
//            var dsKhieuNai = db.KHIEUNAIs
//                .Include("NGUOIDUNG")
//                .Include("SANPHAM")
//                .OrderByDescending(k => k.NgayGui)
//                .ToList();

//            return View(dsKhieuNai);
//        }


//        [HttpPost]
//        public ActionResult CapNhatTrangThaiKN(int id, string trangThai)
//        {
//            var user = Session["user"] as NGUOIDUNG;
//            if (user == null || user.VaiTro != "Admin")
//                return new HttpStatusCodeResult(403, "Không có quyền.");

//            var kn = db.KHIEUNAIs.Find(id);
//            if (kn == null)
//                return HttpNotFound();

//            kn.TrangThai = trangThai;
//            db.SaveChanges();

//            return Json(new { ok = true, status = kn.TrangThai });
//        }

//        // [POST] XÓA SẢN PHẨM TỪ ADMIN
//        [HttpPost]
//        public ActionResult Xoa(int id)
//        {
//            var sp = db.SANPHAMs.Find(id);
//            if (sp == null)
//            {
//                TempData["Error"] = "Sản phẩm không tồn tại!";
//                return RedirectToAction("QuanLySanPham");
//            }

//            try
//            {
//                // 1. Xóa ảnh trên server
//                var hinhAnhs = db.HINHANHSPs.Where(h => h.Masp == id).ToList();
//                string path = Server.MapPath("~/Content/Images/");
//                foreach (var item in hinhAnhs)
//                {
//                    string fullPath = Path.Combine(path, item.URLAnh);
//                    if (System.IO.File.Exists(fullPath) && item.URLAnh != "noimage.jpg")
//                    {
//                        System.IO.File.Delete(fullPath);
//                    }
//                }

//                // 2. Xóa dữ liệu trong DB (Xóa ảnh trước -> Xóa SP sau)
//                db.HINHANHSPs.RemoveRange(hinhAnhs);
//                db.SANPHAMs.Remove(sp);
//                db.SaveChanges();

//                TempData["Success"] = "🗑️ Đã xóa sản phẩm vĩnh viễn!";
//            }
//            catch (Exception)
//            {
//                // Trường hợp lỗi do ràng buộc khóa ngoại (đã có đơn hàng, đánh giá...)
//                // Ta chuyển sang trạng thái Ẩn thay vì xóa
//                sp.TrangThai = "Ẩn";
//                db.SaveChanges();
//                TempData["Success"] = "⚠️ Sản phẩm đã có đơn hàng, không thể xóa hẳn. Hệ thống đã chuyển sang trạng thái 'Ẩn'.";
//            }

//            return RedirectToAction("QuanLySanPham");
//        }
//    }
//}

using Org.BouncyCastle.Asn1.Ocsp;
using System;
using System.Data.Entity;
using System.IO;
using System.Linq;
using System.Web.Mvc;
using ThuongMaiDienTu_DoAn.Filters;
using ThuongMaiDienTu_DoAn.Models;

namespace ThuongMaiDienTu_DoAn.Controllers
{
    [AuthorizeAdmin]
    public class AdminController : Controller
    {
        private readonly TMDTEntities db = new TMDTEntities();

        // === DASHBOARD ===
        public ActionResult Index()
        {
            ViewBag.TongNguoiDung = db.NGUOIDUNGs.Count();
            ViewBag.TongSanPham = db.SANPHAMs.Count();
            ViewBag.Daban = db.SANPHAMs.Count(sp => sp.TrangThai == "Đã bán");
            ViewBag.An = db.SANPHAMs.Count(sp => sp.TrangThai == "Ẩn");
            ViewBag.TinDaDuyet = db.SANPHAMs.Count(sp => sp.TrangThai == "Đã duyệt");
            return View();
        }

        // === DUYỆT TIN / SẢN PHẨM ===


        [HttpPost]
        public ActionResult DoiTrangThai(int id, string tt)
        {
            var sp = db.SANPHAMs.Find(id);
            if (sp == null)
                return HttpNotFound();

            sp.TrangThai = tt;
            db.SaveChanges();

            TempData["Success"] = $"✅ Đã cập nhật trạng thái sản phẩm **{sp.TenSP}** thành '{tt}'";
            return RedirectToAction("QuanLySanPham");
        }

        // === QUẢN LÝ NGƯỜI DÙNG ===
        public ActionResult QuanLyNguoiDung()
        {
            var ds = db.NGUOIDUNGs.OrderByDescending(x => x.NgayTao).ToList();
            return View(ds);
        }

        // === QUẢN LÝ SẢN PHẨM ===
        public ActionResult QuanLySanPham()
        {
            // 1. Lấy dữ liệu Sản phẩm
            var listSP = db.SANPHAMs
                .Include(s => s.NGUOIDUNG)
                .OrderByDescending(x => x.NgayTao)
                .ToList();

            // 2. Lấy dữ liệu Loại (kèm sản phẩm con)
            var listLoai = db.LOAISANPHAMs.Include("SANPHAMs").ToList();

            // 3. Đổ vào ViewModel (lúc này nó tự hiểu class nằm trong thư mục Models)
            var model = new AdminProductViewModel
            {
                SanPhams = listSP,
                LoaiSanPhams = listLoai
            };

            return View(model);
        }

        // === QUẢN LÝ ĐƠN HÀNG ===
        public ActionResult QuanLyDonHang()
        {
            var donhangs = db.HOADONs
                .Select(hd => new DonHangC2CViewModel
                {
                    MaHD = hd.MaHD,
                    NguoiMua = hd.NGUOIDUNG.HoTen,
                    NguoiBan = hd.CT_HOADON
                                 .Select(ct => ct.SANPHAM.NGUOIDUNG.HoTen)
                                 .FirstOrDefault() ?? "(Chưa có sản phẩm)",
                    NgayDat = hd.NgayDat,
                    TongTien = (decimal)hd.TongTien,
                    TrangThai = hd.TrangThai
                })
                .ToList();

            return View(donhangs);
        }

        // === QUẢN LÝ KHIẾU NẠI ===
        public ActionResult QuanLyKhieuNai()
        {
            var dsKhieuNai = db.KHIEUNAIs
                .Include("NGUOIDUNG")
                .Include("SANPHAM")
                .OrderByDescending(k => k.NgayGui)
                .ToList();

            return View(dsKhieuNai);
        }


        [HttpPost]
        public ActionResult CapNhatTrangThaiKN(int id)
        {
            var user = Session["user"] as NGUOIDUNG;
            if (user == null || user.VaiTro != "Admin")
                return new HttpStatusCodeResult(403, "Không có quyền.");

            var kn = db.KHIEUNAIs.Find(id);
            if (kn == null)
                return HttpNotFound();


            kn.TrangThai = "Đã giải quyết";
            kn.PhanHoi = user.HoTen;
            db.SaveChanges();

            //return Json(new { Ma = id,ok = true, responder = kn.PhanHoi, status = kn.TrangThai });
            return RedirectToAction("QuanLyKhieuNai");
        }

        [HttpPost]
        [ValidateAntiForgeryToken]
        public ActionResult XoaKhieuNai(int id)
        {
            var user = Session["user"] as NGUOIDUNG;
            if (user == null || user.VaiTro != "Admin")
                return new HttpStatusCodeResult(403, "Không có quyền.");

            var kn = db.KHIEUNAIs.Find(id);
            if (kn == null)
                return HttpNotFound();

            db.KHIEUNAIs.Remove(kn);
            db.SaveChanges();

            TempData["Success"] = "Đã xoá khiếu nại thành công!";
            return RedirectToAction("QuanLyKhieuNai");
        }

        // [POST] XÓA SẢN PHẨM TỪ ADMIN
        [HttpPost]
        public ActionResult Xoa(int id)
        {
            var sp = db.SANPHAMs.Find(id);
            if (sp == null)
            {
                TempData["Error"] = "Sản phẩm không tồn tại!";
                return RedirectToAction("QuanLySanPham");
            }

            try
            {
                // 1. Xóa ảnh trên server
                var hinhAnhs = db.HINHANHSPs.Where(h => h.Masp == id).ToList();
                string path = Server.MapPath("~/Content/Images/");
                foreach (var item in hinhAnhs)
                {
                    string fullPath = Path.Combine(path, item.URLAnh);
                    if (System.IO.File.Exists(fullPath) && item.URLAnh != "noimage.jpg")
                    {
                        System.IO.File.Delete(fullPath);
                    }
                }

                // 2. Xóa dữ liệu trong DB (Xóa ảnh trước -> Xóa SP sau)
                db.HINHANHSPs.RemoveRange(hinhAnhs);
                db.SANPHAMs.Remove(sp);
                db.SaveChanges();

                TempData["Success"] = "🗑️ Đã xóa sản phẩm vĩnh viễn!";
            }
            catch (Exception)
            {
                // Trường hợp lỗi do ràng buộc khóa ngoại (đã có đơn hàng, đánh giá...)
                // Ta chuyển sang trạng thái Ẩn thay vì xóa
                sp.TrangThai = "Ẩn";
                db.SaveChanges();
                TempData["Success"] = "⚠️ Sản phẩm đã có đơn hàng, không thể xóa hẳn. Hệ thống đã chuyển sang trạng thái 'Ẩn'.";
            }

            return RedirectToAction("QuanLySanPham");
        }
        [HttpPost]
        public ActionResult DoiTrangThaiNguoiDung(int id)
        {
            var user = db.NGUOIDUNGs.FirstOrDefault(x => x.MaKH == id);
            if (user == null)
            {
                TempData["Error"] = "Không tìm thấy tài khoản!";
                return RedirectToAction("QuanLyNguoiDung");
            }

            // Không cho khóa tài khoản Admin
            if (user.VaiTro == "Admin")
            {
                TempData["Error"] = "Không thể khóa tài khoản Admin!";
                return RedirectToAction("QuanLyNguoiDung");
            }

            // Toggle BIT: 0 → 1 hoặc 1 → 0
            user.Khoa = !user.Khoa;

            db.SaveChanges();

            TempData["Success"] = user.Khoa
                ? "Tài khoản đã bị khóa!"
                : "Tài khoản đã được mở khóa!";

            return RedirectToAction("QuanLyNguoiDung");
        }

    }
}

